File: FizzBuzz/java/src/main/java/ProcessJob.java
import java.util.List;

enum ProcessJob {
  $;

  static String process(final int input) throws Exception {
    throwIntegerLessThanOneException(input);
    final List<Filter> filters = FiltersProvider.all();
    final String result = applyFilters(input, filters, "");
    if (result.isEmpty()) {
      return String.valueOf(input);
    }
    return result;
  }

  private static String applyFilters(final int input,
				     final List<Filter> filterers,
				     String result) {
    if (filterers.size() == 0) {
      return result;
    }
    final Filter nextFilter = filterers.remove(0);
    return applyFilters(input, filterers,
			result.concat(nextFilter.filter(input)));
  }

  private static void throwIntegerLessThanOneException(final int input)
      throws Exception {
    if (1 > input) {
	throw new IntegerLessThanOneException(input);
    }
  }
}

File: FizzBuzz/kotlin/src/main/kotlin/ProcessJob.kt
object ProcessJob {
	fun process(input: Int): String
	{
		throwInputLessThanOneException(input)
		val result = applyFilters(input)
		return when (result.isEmpty())
		{
			true -> input.toString()
			else -> result
		}
	}

	private fun applyFilters(input: Int): String
	{
		val filters = FiltersProvider.all()
		return applyFilters(input, filters.toMutableList(), "")
	}

	private tailrec fun applyFilters(input: Int, filters: MutableList<(Int) -> String>, result: String): String
	{
		return when (filters.isEmpty()) {
			true -> result
			else -> {
				val nextFilter = filters.first()
				filters.remove(nextFilter)
				applyFilters(input, filters, result.plus(nextFilter(input)))
			}
		}
	}

	private fun throwInputLessThanOneException(input: Int)
	{
		if (input < 1) {
			throw IntegerLessThanOneException(input)
		}
	}
}
